sklearn/cluster/_k_means.pyx

```diff
@@ -301,13 +311,16 @@ def _centers_dense(np.ndarray[floating, ndim=2] X,
     n_features = X.shape[1]
     cdef int i, j, c
     cdef np.ndarray[floating, ndim=2] centers
-    if floating is float:
-        centers = np.zeros((n_clusters, n_features), dtype=np.float32)
-    else:
-        centers = np.zeros((n_clusters, n_features), dtype=np.float64)
+    cdef np.ndarray[floating, ndim=1] weights_sum_in_cluster
 
-    n_samples_in_cluster = np.bincount(labels, minlength=n_clusters)
-    empty_clusters = np.where(n_samples_in_cluster == 0)[0]
+    dtype = np.float32 if floating is float else np.float64
+    centers = np.zeros((n_clusters, n_features), dtype=dtype)
+    weights_sum_in_cluster = np.zeros((n_clusters,), dtype=dtype)
+
+    for i in range(n_samples):
```

Comment:
Dont know if this is slow since Cython is compiled, yet you may try if `np.add.at` is faster:
```py
weights_sum_in_cluster = np.zeros((n_clusters, ), dtype=dtype)
np.add.at(weights_sum_in_cluster, labels, sample_weights)
empty_clusters = np.where(weights_sum_in_cluster == 0)[0]
```